src/solvers/strings/string_refinement.cpp

```diff
@@ -630,7 +632,18 @@ decision_proceduret::resultt string_refinementt::dec_solve()
 #endif
 
   const union_find_replacet string_id_symbol_resolve =
-    string_identifiers_resolution_from_equations(equations, ns, log.debug());
+    string_identifiers_resolution_from_equations(
+      [&] {
+        std::vector<equal_exprt> equalities;
+        for(const auto &eq : equations)
+        {
+          if(auto equal_expr = expr_try_dynamic_cast<equal_exprt>(eq))
+            equalities.push_back(*equal_expr);
```

Comment:
Not sure I understand. Why do equalities that don't appear as top-level `equal_exprt`s not need to be added to the vector of equalities? :racehorse:

Reply:
If they don't appear at the top then we can't be sure they always hold, for instance if we have `g1 => (s1 == s2)`, we cannot be sure `s1` and `s2` will be equal in the end and we have to keep them as separate symbols (instead of replacing occurences of `s1` by `s2`).